constraint/pkg/client/client.go

```diff
@@ -359,418 +183,402 @@ func (c *Client) AddTemplate(templ *templates.ConstraintTemplate) (*types.Respon
-		return "", fmt.Errorf("%w: empty group for constrant %q",
-			crds.ErrInvalidConstraint, constraint.GetName())
-	}
 
-	if gvk.Kind == "" {
-		return "", fmt.Errorf("%w: empty kind for constraint %q",
-			crds.ErrInvalidConstraint, constraint.GetName())
+	template := c.templates[name]
+	if template == nil {
+		return nil, templateNotFound(name)
 	}
 
-	return path.Join(createConstraintGKSubPath(gvk.GroupKind()), constraint.GetName()), nil
-}
-
-// createConstraintGKPath returns the subpath for given a constraint GK.
-func createConstraintGKSubPath(gk schema.GroupKind) string {
-	return "/" + path.Join("cluster", gk.Group, gk.Kind)
-}
-
-// createConstraintGKPath returns the storage path for a given constrain GK: constraints.<target>.cluster.<group>.<kind>.
-func createConstraintGKPath(target string, gk schema.GroupKind) string {
-	return constraintPathMerge(target, createConstraintGKSubPath(gk))
-}
-
-// constraintPathMerge is a shared function for creating constraint paths to
-// ensure uniformity, it is not meant to be called directly.
-func constraintPathMerge(target, subpath string) string {
-	return "/" + path.Join("constraints", target, subpath)
+	return template.getTemplate(), nil
 }
 
 // getTemplateEntry returns the template entry for a given constraint.
-func (c *Client) getTemplateEntry(constraint *unstructured.Unstructured, lock bool) (*templateEntry, error) {
-	kind := constraint.GetKind()
-	if kind == "" {
-		return nil, fmt.Errorf("%w: kind missing from Constraint %q",
-			crds.ErrInvalidConstraint, constraint.GetName())
-	}
-
-	group := constraint.GroupVersionKind().Group
-	if group != constraints.Group {
-		return nil, fmt.Errorf("%w: wrong API Group for Constraint %q, got %q but need %q",
-			crds.ErrInvalidConstraint, constraint.GetName(), group, constraints.Group)
-	}
-
-	if lock {
-		c.mtx.RLock()
-		defer c.mtx.RUnlock()
-	}
+func (c *Client) getTemplateForKind(kind string) *templateClient {
+	name := strings.ToLower(kind)
 
-	entry, ok := c.templates[templateKeyFromConstraint(constraint)]
-	if !ok {
-		var known []string
-		for k := range c.templates {
-			known = append(known, string(k))
-		}
-
-		return nil, fmt.Errorf("%w: Constraint kind %q is not recognized, known kinds %v",
-			ErrMissingConstraintTemplate, kind, known)
-	}
-
-	return entry, nil
+	return c.templates[name]
 }
 
 // AddConstraint validates the constraint and, if valid, inserts it into OPA.
 // On error, the responses return value will still be populated so that
 // partial results can be analyzed.
 func (c *Client) AddConstraint(ctx context.Context, constraint *unstructured.Unstructured) (*types.Responses, error) {
+	resp := types.NewResponses()
+
 	c.mtx.Lock()
 	defer c.mtx.Unlock()
 
-	resp := types.NewResponses()
-	entry, err := c.getTemplateEntry(constraint, false)
+	err := c.validateConstraint(constraint)
 	if err != nil {
 		return resp, err
 	}
 
-	subPath, err := createConstraintSubPath(constraint)
+	kind := constraint.GetKind()
```

Comment:
check for `kind != ""`

Reply:
We do this in `validateConstraintMetadata` just before, and this code path has a unit test.